File: vscode/src/core/CoordinatorState.ts
export interface CoordinatorRequest<T> {
  editorId: T;
  generation: number;
}

export class CoordinatorState<T> {
  private activeEditorId: T | null = null;
  private windowFocused = false;
  private enabled = true;
  private generation = 0;
  private shuttingDown = false;

  focusEditor(editorId: T): void {
    if (this.shuttingDown) return;
    if (this.windowFocused && this.activeEditorId === editorId) return;
    this.activeEditorId = editorId;
    this.windowFocused = true;
    this.generation += 1;
  }

  loseFocus(editorId?: T): void {
    if (editorId !== undefined && this.activeEditorId !== editorId) return;
    this.activeEditorId = null;
    this.windowFocused = false;
    this.generation += 1;
  }

  setEnabled(value: boolean): void {
    if (this.enabled === value) return;
    this.enabled = value;
    if (!value) {
      this.activeEditorId = null;
    }
    this.generation += 1;
  }

  newRequest(editorId: T): CoordinatorRequest<T> | null {
    if (
      this.shuttingDown ||
      !this.enabled ||
      !this.windowFocused ||
      this.activeEditorId !== editorId
    ) {
      return null;
    }
    this.generation += 1;
    return { editorId, generation: this.generation };
  }

  invalidateRequests(): void {
    this.generation += 1;
  }

  isCurrent(request: CoordinatorRequest<T>): boolean {
    return (
      !this.shuttingDown &&
      this.enabled &&
      this.windowFocused &&
      this.activeEditorId === request.editorId &&
      this.generation === request.generation
    );
  }

  shutdown(): void {
    if (this.shuttingDown) return;
    this.shuttingDown = true;
    this.activeEditorId = null;
    this.windowFocused = false;
    this.generation += 1;
  }

  isShuttingDown(): boolean {
    return this.shuttingDown;
  }
}

File: core/src/main/kotlin/com/auto_switch_ime/core/coordinator/CoordinatorState.kt
package com.auto_switch_ime.core.coordinator

data class CoordinatorRequest<T>(
    val editorId: T,
    val generation: Long
)

class CoordinatorState<T> {
    private var activeEditorId: T? = null
    private var windowFocused = false
    private var enabled = true
    private var generation = 0L
    private var shuttingDown = false

    @Synchronized
    fun focusEditor(editorId: T) {
        if (shuttingDown) return
        if (windowFocused && activeEditorId == editorId) return
        activeEditorId = editorId
        windowFocused = true
        generation++
    }

    @Synchronized
    fun loseFocus(): Boolean {
        clearFocus()
        return true
    }

    @Synchronized
    fun loseFocus(editorId: T): Boolean {
        if (activeEditorId != editorId) return false
        clearFocus()
        return true
    }

    private fun clearFocus() {
        activeEditorId = null
        windowFocused = false
        generation++
    }

    @Synchronized
    fun setEnabled(value: Boolean) {
        if (enabled == value) return
        enabled = value
        if (!value) {
            activeEditorId = null
        }
        generation++
    }

    @Synchronized
    fun newRequest(editorId: T): CoordinatorRequest<T>? {
        if (shuttingDown || !enabled || !windowFocused || activeEditorId != editorId) {
            return null
        }
        generation++
        return CoordinatorRequest(editorId, generation)
    }

    @Synchronized
    fun invalidateRequests() {
        generation++
    }

    @Synchronized
    fun isCurrent(request: CoordinatorRequest<T>, platformFocused: Boolean): Boolean {
        return !shuttingDown &&
                enabled &&
                windowFocused &&
                platformFocused &&
                activeEditorId == request.editorId &&
                generation == request.generation
    }

    @Synchronized
    fun shutdown() {
        if (shuttingDown) return
        shuttingDown = true
        activeEditorId = null
        windowFocused = false
        generation++
    }

    @Synchronized
    fun isShuttingDown(): Boolean = shuttingDown
}
